pkg/credentialprovider/shim/shim.go

```diff
@@ -0,0 +1,189 @@
+			// which containerd will try first against the configured mirror (which should fail as incorrect credentials for
+			// this registry) and then against the origin registry.
+			authMap[globbedDomain] = mirrorAuth
+		case v1alpha1.MirrorCredentialsFirst:
+			// Set mirror credentials for image to ensure that the mirror credentials are used first, and set any existing
+			// origin credentials for the globbed domain to ensure they are used last (glob matches have lowest precedence).
+			//
+			// This means that the kubelet will first try the origin credentials, which containerd will try against both the
+			// configured mirror in containerd (which should fail as incorrect credentials for this registry) and the origin
+			// registry.
+			//
+			// If containerd fails to pull using the origin credentials, then the kubelet will try the mirror credentials,
+			// which containerd will try first against the configured mirror and then against the origin registry (which
+			// should fail as incorrect credentials for this registry) if the image is not found in the mirror.
+			existing, found := authMap[img]
+			if found {
+				authMap[globbedDomain] = existing
+			}
+			authMap[img] = mirrorAuth
+		default:
+			return nil, fmt.Errorf(
+				"%w: %q",
+				ErrUnsupportedMirrorCredentialStrategy,
+				p.cfg.Mirror.MirrorCredentialsStrategy,
+			)
+		}
+	}
+
+	return &credentialproviderv1beta1.CredentialProviderResponse{
+		CacheKeyType:  credentialproviderv1beta1.ImagePluginCacheKeyType,
+		CacheDuration: &metav1.Duration{Duration: 0},
+		Auth:          authMap,
+	}, nil
+}
+
+func (p shimProvider) getMirrorCredentials(
+	img string, //nolint:unparam,revive // Placeholder for now.
+) (credentialproviderv1beta1.AuthConfig, bool) { //nolint:unparam // Placeholder for now
+	// If mirror is not configured then return no credentials for the mirror.
+	if p.cfg.Mirror == nil {
+		return credentialproviderv1beta1.AuthConfig{}, false
+	}
+
+	// TODO Call relevant credential provider plugin based on the image domain replaced with the mirror URL to get the
+	// credentials for the mirror.
+
+	return credentialproviderv1beta1.AuthConfig{}, false
+}
+
+func (p shimProvider) getOriginCredentials(
+	img string, //nolint:unparam,revive // Placeholder for now
+) (credentialproviderv1beta1.AuthConfig, bool) { //nolint:unparam // Placeholder for now
+	// If only mirror credentials should be used then return no credentials for the origin.
+	if isRegistryCredentialsOnly(p.cfg.Mirror) {
+		return credentialproviderv1beta1.AuthConfig{}, false
+	}
+
+	// TODO Call relevant credential provider plugin based on the image to get the credentials for the origin.
+
+	return credentialproviderv1beta1.AuthConfig{}, false
+}
+
+func isRegistryCredentialsOnly(cfg *v1alpha1.MirrorConfig) bool {
+	return cfg != nil &&
+		cfg.MirrorCredentialsStrategy == v1alpha1.MirrorCredentialsOnly
+}
+
+func globbedDomainForImage(img string) (string, error) {
```

Comment:
This looks right, but some unit tests would be great 🙏

Reply:
Copied some stuff from upstream, moved globber to own package, and added unit tests.